### PR TITLE
Fix some issues with lineWidth < 1 after transform (bug 1753075, bug 1743245, bug 1710019)

### LINKED ISSUE #13211: v2.7-2.8 incorrectly render dashed line in FPDF/FPDI-created files.
Attach (recommended) or Link to PDF file here:
https://kanuay.com/pdf/report/test.pdf

Configuration:
- Web browser and its version: Firefox Dev 88.0 & Safari iOS 14
- Operating system and its version: Windows 10 Latest & iOS 14
- PDF.js version: 2.7-2.8 (Both self-hosted and Firefox built-in)

Steps to reproduce the problem:
Lot of my file created by FPDF/FPDI
Open these files with PDF.js v2.7 or v2.8
All the fraction lines, radical signs, especially dashed lines are rendered incorrectly.

PDF js Problem Report 210409

### LINKED ISSUE #14521: PDF contents turns redacted (blackout) upon zoom-out to thumbnail size.
Attach (recommended) or Link to PDF file here:

Sample Document : 
[Sample_blackout.pdf](https://github.com/mozilla/pdf.js/files/7977687/Sample_blackout.pdf)

Configuration:
- Web browser and its version: Chrome, Firefox, MS Edge (latest versions)
- Operating system and its version: Windows, Mac
- PDF.js version: 2.12.313 (latest)
- Is a browser extension: NA

Steps to reproduce the problem:
1. Open attached `Sample_blackout.pdf` in pdfjs viewer
2. Zoom out the document to lesser than 30% or verify the content in thumbnail view
3. You will notice portion of the document content are turned black.

What is the expected behavior? : Document content should not change its style in thumbs view or while zoom out. 

What went wrong? : document content are turned black and even some portions are redacted.

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): https://mozilla.github.io/pdf.js/web/viewer.html

**Output in pdfjs viewer on zoom out :** 

blackedOutResponse

### PR DESCRIPTION
**Pull request by calixteman**
- it aims to fix:
   - https://bugzilla.mozilla.org/show_bug.cgi?id=1753075;
   - https://bugzilla.mozilla.org/show_bug.cgi?id=1743245;
   - https://bugzilla.mozilla.org/show_bug.cgi?id=1710019;
   - issue #13211;
   - issue #14521. 
 - previously we were trying to adjust lineWidth to have something correct after the current transform is applied but this approach was not correct because finally the pixel is rescaled with the same factors in both directions.
  And sometimes those factors must be different (see bug 1753075).
 - So the idea of this patch is to apply a scale matrix to the current transform just before setting lineWidth and stroking. This scale matrix is computed in order to ensure that after transform, a pixel will have its two thickness greater than 1.

### CODE REVIEW

**Comment by marco-c** [comment]
Is this fixing also the bugs linked to bug 1753075 ([1743245](https://bugzilla.mozilla.org/show_bug.cgi?id=1743245) and [1657838](https://bugzilla.mozilla.org/show_bug.cgi?id=1657838)) and the issues linked to #13211 (#14521)?
If yes, we might want to add the PDFs from those bugs/issues to the reftests too.

**Review by Snuffleupagus — COMMENTED** [review]
Hopefully @brendandahl can review this, since I don't feel that I know all of this code well enough to do so properly myself.

**Comment by Snuffleupagus** [comment]
Besides the regressions related to text-stroking, there also seem to be some regressions related to *dashed* lines in a number of test-cases (e.g. even `tracemonkey-eq-page2` regressed a bit).

**Comment by calixteman** [comment]
Yeah I noticed the ones about the dashed lines but I don't see them locally.

**Comment by Snuffleupagus** [comment]
> Yeah I noticed the ones about the dashed lines but I don't see them locally.

I've *just* tested this patch locally on Windows, using a regular screen (no hiDPI), and I can reproduce this directly in the viewer at various zoom levels.

File: src/display/canvas.js [hunk 1]
```diff
@@ -1419,6 +1416,9 @@ class CanvasGraphics {
 
   // Graphics state
   setLineWidth(width) {
+    if (width !== this.current.lineWidth) {
+      this._cachedScaleForStroking = null;
```
Comment (Snuffleupagus) [inline src/display/canvas.js:1420]: Is it not necessary to reset *both* caches here?

**Comment by marco-c** [comment]
> Is this fixing also the bugs linked to bug 1753075 ([1743245](https://bugzilla.mozilla.org/show_bug.cgi?id=1743245) and [1657838](https://bugzilla.mozilla.org/show_bug.cgi?id=1657838)) and the issues linked to #13211 (#14521)? If yes, we might want to add the PDFs from those bugs/issues to the reftests too.

@calixteman ping

File: src/display/canvas.js (hunk 1 above)
Comment (calixteman) [reply src/display/canvas.js:1420]: The other cached value doesn't use the lineWidth, so it's useless.

File: src/display/canvas.js [hunk 2]
```diff
@@ -3104,46 +3069,140 @@ class CanvasGraphics {
...
+    }
+
+    const savedMatrix = ctx.mozCurrentTransform.slice();
+    const savedDashes = ctx.getLineDash().slice();
+    const savedDashOffset = ctx.lineDashOffset;
+
+    ctx.scale(scaleX, scaleY);
+    ctx.lineWidth = lineWidth || 1;
+
+    // How the dashed line is rendered depends on the current transform...
```
Comment (brendandahl) [inline src/display/canvas.js:3168]: Can we make this chunk instead call rescaleAndStrokeNoSave or add another function that they both call. Having two code paths to keep in sync can be problematic.

**Comment by brendandahl** [comment]
On a high dpi display the lines feel very "heavy" compared to the previous code. 
New code vs acrobat
image
New code vs acrobat
image
Old code vs acrobat
image

**Comment by Snuffleupagus** [comment]
Note that a bunch of tests failed with `.... | render : TypeError: this.rescaleAndStrokeNoSave is not a function`, since that method doesn't exist in the last version of the patch :-P

**Comment by calixteman** [comment]
@Snuffleupagus merci
/botio test

**Pull request merged**